projects/opendr_ws_2/src/opendr_perception/opendr_perception/object_tracking_2d_deep_sort_node.py

```diff
@@ -0,0 +1,246 @@
+        :type device: str
+        :param model_name: the pretrained model to download or a saved model in temp_dir folder to use
+        :type model_name: str
+        :param temp_dir: the folder to download models
+        :type temp_dir: str
+        """
+
+        super().__init__('object_tracking_2d_deep_sort_node')
+
+        self.get_logger().info("Using model_name: {}".format(model_name))
+
+        self.detector = detector
+        self.learner = ObjectTracking2DDeepSortLearner(
+            device=device, temp_path=temp_dir,
+        )
+        if not os.path.exists(os.path.join(temp_dir, model_name)):
+            ObjectTracking2DDeepSortLearner.download(model_name, temp_dir)
+
+        self.learner.load(os.path.join(temp_dir, model_name), verbose=True)
+
+        # Initialize OpenDR ROSBridge object
+        self.bridge = ROS2Bridge()
+        self.tracking_id_publisher = self.create_publisher(
+            Int32MultiArray, output_tracking_id_topic, 1
+        )
+
+        if output_image_topic is not None:
+            self.output_image_publisher = self.create_publisher(
+                ROS_Image, output_image_topic, 1
+            )
+
+        self.detection_publisher = self.create_publisher(
+            Detection2DArray, output_detection_topic, 1
+        )
+
+        self.create_subscription(ROS_Image, input_image_topic, self.callback, 1)
+
+    def callback(self, data):
+        """
+        Callback that process the input data and publishes to the corresponding topics
+        :param data: input message
+        :type data: sensor_msgs.msg.Image
+        """
+
+        # Convert sensor_msgs.msg.Image into OpenDR Image
+        image = self.bridge.from_ros_image(data, encoding="bgr8")
+        detection_boxes = self.detector.infer(image)
+        image_with_detections = ImageWithDetections(image.numpy(), detection_boxes)
+        tracking_boxes = self.learner.infer(image_with_detections)
+
+        if self.output_image_publisher is not None:
+            frame = image.opencv()
+            draw_predictions(frame, tracking_boxes)
+            message = self.bridge.to_ros_image(
+                Image(frame), encoding="bgr8"
+            )
+            self.output_image_publisher.publish(message)
+            self.get_logger().info("Published annotated image")
+
+        ids = [int(tracking_box.id) for tracking_box in tracking_boxes]
+
+        # Convert detected boxes to ROS type and publish
+        ros_boxes = self.bridge.to_ros_boxes(detection_boxes)
+        if self.detection_publisher is not None:
+            self.detection_publisher.publish(ros_boxes)
+            self.get_logger().info("Published " + str(len(detection_boxes)) + " detection boxes")
+
+        ros_ids = Int32MultiArray()
+        ros_ids.data = ids
+
+        if self.tracking_id_publisher is not None:
+            self.tracking_id_publisher.publish(ros_ids)
+            self.get_logger().info("Published " + str(len(ids)) + " tracking ids")
```

Comment:
Moved around a few lines of code as a minor optimization in case a user disables one of the outputs.

```suggestion

        # Convert detected boxes to ROS type and publish
        if self.detection_publisher is not None:
            ros_boxes = self.bridge.to_ros_boxes(detection_boxes)
            self.detection_publisher.publish(ros_boxes)
            self.get_logger().info("Published " + str(len(detection_boxes)) + " detection boxes")

        if self.tracking_id_publisher is not None:
            ids = [int(tracking_box.id) for tracking_box in tracking_boxes]
            ros_ids = Int32MultiArray()
            ros_ids.data = ids
            self.tracking_id_publisher.publish(ros_ids)
            self.get_logger().info("Published " + str(len(ids)) + " tracking ids")
```